clang/lib/Sema/SemaDecl.cpp

```diff
@@ -12222,8 +12222,17 @@ bool Sema::CheckFunctionDeclaration(Scope *S, FunctionDecl *NewFD,
             << NewFD;
     }
 
-    if (!Redeclaration && LangOpts.CUDA)
+    if (!Redeclaration && LangOpts.CUDA) {
```

Comment:
Is it possible to add a check for sm70+ for grid_constant here?

Reply:
I deliberately decided *not* to do that.
`__grid_constant__` is an optimization hint, and LLVM *will* ignore it if the attribute is present, but we're targeting an older GPU, or generating an older PTX version. Considering that using the attribute with pre-sm70 GPU does no harm I do not think it should be an error.

NVCC's insistence on erroring out if the attribute is used during compilation for an older GPU is  unnecessaryand it makes the attribute a pain to use in practice. Considering that it's not uncommon to compile the same source code for multiple GPUs, including the older ones, we'll have to resort to preprocessor tweaks and undef the attribute way, if we want to give compiler a hint that we want to avoid copying some arguments.

Reply:
This makes sense. Perhaps a warning? But even that might not be necessary- anyone who is using `__grid_constant__` knows what they are using it for.

Reply:
Makes sense. I also realized that you'd already addressed this in the description 🙂 
> One notable difference from NVCC is that the attribute can be used regardless of the targeted GPU. On the older GPUs it will just be ignored. The attribute is a performance hint, and does not warrant a hard error if compiler can't benefit from it on a particular GPU variant.